wip-trampoline.md

```diff
@@ -0,0 +1,138 @@
+trampoline hop's address, transfer information, and its associated HMAC.
+
+It is `2*N*FRAME_SIZE` long and has the following structure (where `n1` and
+`n2` are the number of frames used by trampoline hops `1` and `2`):
+
+1. type: `trampoline_hops_data`
+2. data:
+   * [`n1*FRAME_SIZE`:`trampoline_hop_payload`]
+   * [`n2*FRAME_SIZE`:`trampoline_hop_payload`]
+   * ...
+   * `filler`
+
+Each `trampoline_hop_payload` has the following structure:
+
+1. type: `trampoline_hop_payload`
+2. data:
+   * [`1`:`num_frames_and_realm`]
+   * [`1`:`type`]
+   * [`1`:`length`]
+   * [`33`:`node_id`]
+   * [`8`:`amt_to_forward`]
+   * [`4`:`outgoing_cltv_value`]
+   * `optional TLV can be added here`
+   * `padding`
+   * [`32`:`HMAC`]
+
+For this specification, some restrictions apply:
+
+* `realm` MUST be `0x01` to indicate that the payload contains a TLV stream
+and `num_frames` MUST correctly reflect the number of frames used.
+* the TLV stream MUST contain the `type` `0x01` (`next_trampoline_hop`)
+* `next_trampoline_hop`'s length MUST be `33+8+4=45` bytes
+* the TLV stream MAY contain other optional types
+* `padding` MUST use only `0x00` bytes
+
+TODO: it's very sad to use 2 frames and have 50 bytes wasted per hop...
+But that can also be seen as an opportunity to add optional data to help routing ¯\_(ツ)_/¯
+Let's try to improve this though it would be great to be able to use a single
+frame per trampoline...
+
+### Nesting a Trampoline Onion in a v0 onion
+
+Trampoline onion packets combined with multi-frame v0 onion packets make it
+possible for nodes to route payments without a complete view of the network.
+
+The sending node (and each trampoline hop) uses TLV type `0x01` inside a v0
+onion packet to send a `trampoline_onion_packet` as last `hop_payload`:
+
+1. type: `onion_packet`
+2. data:
+   * [`1`:`version`]
+   * [`33`:`public_key`]
+   * [`n1*FRAME_SIZE`:`hop_payload`]
+   * [`n2*FRAME_SIZE`:`hop_payload`]
+   * ...
+   * [`2*(N+1)*FRAME_SIZE`:`trampoline_onion_packet_payload`]
+   * `filler`
+   * [`32`:`hmac`]
+
+The structure of the last payload is:
+
+1. type: `trampoline_onion_packet_payload`
+2. data:
+   * [`1`:`realm`]
+   * [`1`:`type`]
+   * [`3`:`length`]
+   * [`2*N*FRAME_SIZE + 66`:`trampoline_onion_packet`]
+   * `optional TLV can be added here`
+   * `padding`
+
+The maximum number of v0 hops is `20-2*(N+1)`.
+
+TODO: here as well we waste a whole frame for the trampoline onion version, TLV
+type and TLV length (which is only 5 bytes). This is very wasteful and can be
+improved.
+
+Upon receiving a packet with type `trampoline_onion_packet_payload`:
+
+* The receiving node is the destination for this v0 onion packet
+* It extracts the `trampoline_onion_packet` and is able to unwrap one layer to
+discover the next trampoline node
+* It creates a new v0 onion packet to the next trampoline node
+* It puts the unwrapped `trampoline_onion_packet` in the payload for the last
+hop
```

Comment:
Notice that we also need a downgrade option: while we may want to rely on trampolines to forward a payment through the network and each of the trampolines just receives an inner onion it knows how to process, the recipient may not support trampolines at all. Hence the last trampoline may be asked to downgrade to a regular `hop_payload` so that the onion arriving at the recipient is just a classical onion.

Reply:
While this seems to be addressed in the example below I can't find a description of how this should really work.

Reply:
Very good point, for now I assumed that the destination would need to handle trampoline (and ideally provide an invoice with hints about close trampoline nodes).
However you're right that we should make this work for legacy payee nodes. I'll figure something out.